### PR TITLE
[nextercism] Add user config tests for Windows

### PR DESCRIPTION
**Pull request by robphoenix**
I've ported the darwin tests to Windows, which led to a couple of changes in the `user_config.go` file due to different assumptions on Windows, and I've moved the test in `user_config_test.go` file to the darwin and windows tests suites because of these different assumptions.

### CODE REVIEW

File: config/user_config.go [hunk 1]
```diff
@@ -84,9 +84,18 @@ func (cfg *UserConfig) resolve(path string) string {
 	if strings.HasPrefix(path, "~"+string(os.PathSeparator)) {
 		return strings.Replace(path, "~", cfg.Home, 1)
 	}
+	// if using "~/" on Windows
```
Comment (robphoenix) [inline config/user_config.go]: On Windows it's entirely plausible to write `~/dir` to indicate `C:\\Users\\alice\\dir`, yet `os.PathSeperator` on Windows is `\` so the conditional above doesn't work, but it does if someone writes `~\\dir`

File: config/user_config.go [hunk 2]
```diff
@@ -84,9 +84,18 @@ func (cfg *UserConfig) resolve(path string) string {
...
+	if strings.HasPrefix(path, "~/") {
+		path = strings.Replace(path, "~/", "", 1)
+		return filepath.Join(cfg.Home, path)
+	}
 	if filepath.IsAbs(path) {
 		return filepath.Clean(path)
 	}
+	// if using "/dir" on Windows
```
Comment (robphoenix) [inline config/user_config.go]: On linux/darwin `/dir` represents a directory in the root, which on windows would be `C:\\`, yet my assumption would be that if a Windows user meant to use the C drive they would write `C:\\dir` and that writing `/dir` would indicate a relative dir, relative to `cfg.Home`. I may be wrong in this assumption, so happy to hear if there could be a better way.

File: config/user_config_test.go [hunk 3]
```diff
@@ -1,35 +0,0 @@
...
-	assert.NoError(t, err)
-
-	// reload it
-	cfg = &UserConfig{
-		Config: New(dir, "user"),
-	}
-	err = cfg.Load(viper.New())
-	assert.NoError(t, err)
-	assert.Equal(t, "a", cfg.Token)
-	assert.Equal(t, "/a", cfg.Workspace)
```
Comment (robphoenix) [inline config/user_config_test.go]: I couldn't find a way to make this work across *nix & Windows, based on the different assumption of what `/a` means.

File: config/user_config_windows_test.go [hunk 4]
```diff
@@ -1,9 +1,62 @@
...
+	assert.NoError(t, err)
+
+	// reload it
+	cfg = &UserConfig{
+		Config: New(dir, "user"),
+	}
+	err = cfg.Load(viper.New())
+	assert.NoError(t, err)
+	assert.Equal(t, "a", cfg.Token)
+	assert.Equal(t, filepath.Join(cfg.Home, "a"), cfg.Workspace)
```
Comment (robphoenix) [inline config/user_config_windows_test.go]: Here, again, the assumption that `/a` would mean `C:\\Users\\alice\\a`

File: config/user_config_windows_test.go [hunk 5]
```diff
@@ -1,9 +1,62 @@
...
+	cwd, err := os.Getwd()
+	assert.NoError(t, err)
+
+	cfg := &UserConfig{Home: "C:\\Users\\alice"}
+	tests := []struct {
+		in, out string
+	}{
+		{"", ""}, // don't make wild guesses
+		{"C:\\Users\\alice\\\\\\foobar", "C:\\Users\\alice\\foobar"},
+		{"/foobar", "C:\\Users\\alice\\foobar"},
```
Comment (robphoenix) [inline config/user_config_windows_test.go]: An extra test from the Darwin ones for assuming that `/foobar` is in the home directory.

File: config/user_config_test.go (hunk 3 above)
Comment (kytrinyx) [inline config/user_config_test.go]: Ah, yeah. We don't even need to have a `/a` just some jumble that comes back out the same.

File: config/user_config_windows_test.go (hunk 4 above)
Comment (kytrinyx) [inline config/user_config_windows_test.go]: This works for me.

File: config/user_config.go [hunk 6]
```diff
@@ -84,9 +84,18 @@ func (cfg *UserConfig) resolve(path string) string {
 	if strings.HasPrefix(path, "~"+string(os.PathSeparator)) {
 		return strings.Replace(path, "~", cfg.Home, 1)
 	}
+	// if using "~/" on Windows
+	if strings.HasPrefix(path, "~/") {
+		path = strings.Replace(path, "~/", "", 1)
+		return filepath.Join(cfg.Home, path)
+	}
```
Comment (kytrinyx) [inline config/user_config.go]: I think if we have this, then we can delete lines 84-86, as your version should work on all platforms, right?

File: config/user_config_test.go (hunk 3 above)
Comment (robphoenix) [inline config/user_config_test.go]: Ok, shall I move the test back and fix it?

File: config/user_config.go (hunk 6 above)
Comment (robphoenix) [inline config/user_config.go]: The only case that would fail is a Windows user typing `~\\dir`

File: config/user_config_test.go (hunk 3 above)
Comment (robphoenix) [inline config/user_config_test.go]: hang on, it's broken, I need my Linux.

**Comment by robphoenix** [comment]
So I've refactored the test suite a bit, I may be way off, so do say if it's off the mark, and don't feel this needs to be anywhere near the top of your todo list, I imagine it's fairly substantial!

Initially I found a way to make the main user config test pass on Windows and *nix, then I copied the darwin tests to a linux test file and I didn't like having two test files that were exactly the same, so I thought we might be better off having the main `user_config_test.go` file that builds for linux and darwin, and then because it deals with filepath's so much, a separate Windows test file that can veer off and deal with Windows idiosyncrasies, without trying to make them play nice together. Does that make sense? Sorry, rambling.

I wonder if we need to set up appveyor, for these Windows tests.

File: config/user_config_test.go (hunk 3 above)
Comment (kytrinyx) [inline config/user_config_test.go]: > I need my Linux.

Don't we all 😂

**Comment by kytrinyx** [comment]
> Does that make sense?

Yeah that makes a lot of sense. Let me take a look through the whole thing, but your reasoning certainly sounds good to me.

> I wonder if we need to set up appveyor, for these Windows tests.

Yes! It's time. I'll get that set up. (Should be super straight forward, I already have an account).

For the build constraint... I suddenly wondered if we need a space between the constraint and the package declaration. Do you have your windows laptop handy to check if that would work? (Maybe AppVeyor will find out in a few minutes).

File: config/user_config_windows_test.go [hunk 7]
```diff
@@ -1,9 +1,61 @@
...
-	t.Fatal("Need to implement normalization tests for windows.")
+	cwd, err := os.Getwd()
+	assert.NoError(t, err)
+
+	cfg := &UserConfig{Home: "C:\\Users\\alice"}
+	tests := []struct {
+		in, out string
+	}{
+		{"", ""}, // don't make wild guesses
+		{"C:\\Users\\alice\\\\\\foobar", "C:\\Users\\alice\\foobar"},
```
Comment (kytrinyx) [inline config/user_config_windows_test.go]: Thanks for cleaning this up. Did you see #406? Do you know if this normalization will fix that?

**Comment by robphoenix** [comment]
> I suddenly wondered if we need a space between the constraint and the package declaration.

:bulb: Yes! that's it, I'll open a PR now.

File: config/user_config_windows_test.go (hunk 7 above)
Comment (robphoenix) [inline config/user_config_windows_test.go]: I'll check and add some tests to cover those situations

File: config/user_config_test.go (hunk 3 above)
Comment (robphoenix) [inline config/user_config_test.go]: I blame @jessfraz

Comment (kytrinyx) [inline config/user_config_test.go]: LOL. As you should :-)

File: config/user_config_windows_test.go (hunk 7 above)
Comment (robphoenix) [inline config/user_config_windows_test.go]: As far as I can tell it should. In the tests we have to have `\\` to signify `\`, the first one to escape the second, as I'm sure you realise. So here `C:\\Users\\alice` would be the same as typing `C:\Users\alice` at the command line.

Just to be sure I'll build this and give it a whirl tomorrow, doing what's mentioned in #406

**Comment by kytrinyx** [comment]
@robphoenix OK if I add you to the committers on this repo?

@nywilken same goes for you.

I have you both as a generic "contributor" so that stuff can be assigned, but I think both of you have been doing great work both here and in Configlet, and I'd love to make it official here as well.

**Comment by robphoenix** [comment]
> Oh. Wait.

Yeah, sorry, jumping from one machine to the other, not doing my git properly. Should be all good to merge now, phew.

> OK if I add you to the committers on this repo?

Absolutely! Thankyou :pray: :heart:

**Comment by robphoenix** [comment]
> When you're happy with it go ahead and merge.

I don't have the permissions to merge yet.

**Comment by nywilken** [comment]
>@robphoenix OK if I add you to the committers on this repo?
> @nywilken same goes for you.

Absolutely! Thank you.

File: config/user_config_windows_test.go (hunk 7 above)
Comment (kytrinyx) [inline config/user_config_windows_test.go]: Yeah, I'm just not sure how we ended up adding slashes in the current one /shrug

**Comment by kytrinyx** [comment]
I've added you both—I'll let one of you do the honors :)

**Pull request merged**